base/reflection.jl

```diff
@@ -1778,6 +1778,79 @@ function return_types(@nospecialize(f), @nospecialize(types=default_tt(f));
     return rts
 end
 
+"""
+    Base.return_type(
+        f, types=default_tt(f);
+        world::UInt=get_world_counter(),
+        interp::Core.Compiler.AbstractInterpreter=Core.Compiler.NativeInterpreter(world)) -> rt::Type
+
+Returns the return type of the function call specified by `f` and `types`.
```

Comment:
Probably should say "a supertype of the return type"

And seems like `Base.infer_return_type` would be consistent with the names of other similar queries?

Also might want to add a See also ref to `Base.promote_op` & friends so that people get directed to ways to infer these for generated functions, rather than just being warned that using from a generated function is an error

Reply:
> Probably should say "a supertype of the return type"

Maybe "an inferred return type" would be better? "supertype" sounds like returning something that is strictly wider in the type lattice. "inferred" might be a bit ambiguous, but it can still effectively communicate the idea.

Reply:
> And seems like `Base.infer_return_type` would be consistent with the names of other similar queries?

Yeah. `Base.infer_return_type` isn't consistent with `Base.return_types`, but we might want to rename `Base.return_types` to `Base.infer_return_types` instead.

Reply:
> Also might want to add a See also ref to `Base.promote_op` & friends so that people get directed to ways to infer these for generated functions, rather than just being warned that using from a generated function is an error

Are you suggesting that when it's absolutely necessary to infer the return type from a generated context, we should lean towards using `Core.Compiler.return_type`? Or is the point more about using the docstring of `promote_op` as a means to highlight the nuances of using these reflection tools?

Reply:
It is never allowed directly, but you can pass it as an argument to the generated function

Reply:
Do you mean the implicit `world::UInt` argument that is passed to the generator function? I'm not sure how `promote_op` relates to generated functions.

Reply:
no, just as a real argument